sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala

```diff
@@ -77,4 +78,42 @@ private[sql] object StatFunctions {
       })
     counts.cov
   }
+
+  /** Generate a table of frequencies for the elements of two columns. */
+  private[sql] def crossTabulate(df: DataFrame, col1: String, col2: String): DataFrame = {
+    val tableName = s"${col1}_$col2"
+    val distinctCol2 = df.select(col2).distinct.collect().sortBy(_.get(0).toString)
```

Comment:
Interesting, this comment only shows up in the diff. +1 on the single-pass approach. Driver could be the bottleneck, but we are not expecting large amount of data for `crosstab`.